### ISSUE TITLE
Assertion failed: getNumParams() == params.size() && "NumParams overflow!" 

### ISSUE THREAD

**Issue by corwinn** [llvm/llvm-project#60047]
Code: https://gist.github.com/corwinn/f2c006299db2ff3c9d36fe36cf263cbe
Compile: clang++ -std=c++14  variadic_template_issue.cpp
331334 bytes error message, containing: "clang frontend command failed due to signal"
The code is one line, the arch and platform shouldn't matter.

Regards,
Corwin.

**Comment by tbaederr** [comment]
https://godbolt.org/z/jseGEzP8v

**Comment by shafik** [comment]
This is infinite instantiation exhausting system resources, MSVC and gcc also crash on this code: https://godbolt.org/z/sb1jzPGWd

It would be great to catch this and emit a diagnostic though

@AaronBallman do we catch any similar cases of resource exhaustion in other cases?

**Comment by AaronBallman** [comment]
> @AaronBallman do we catch any similar cases of resource exhaustion in other cases?

Yup! We have `runWithSufficientStackSpace()` for this sort of thing. As an example: https://github.com/llvm/llvm-project/blob/main/clang/lib/Parse/ParseDecl.cpp#L6118
